internal/clients/telegram/client.go

```diff
@@ -264,231 +309,171 @@ However it is not as simple as that. After an update is processed it could chang
-this case /start) would be the answer. If we process /quiz and /start at the same time /start update would have no idea
-that it should be handled as an answer to /quiz, and the user will get 2 messages: the first one is the quiz question
-and the second one is the greeting message from /start. Then the user tries to answer the quiz and nothing happens
-because /start saved the state last and overwrote what /quiz did.
-
-TODO: If the state for a conversation was given out and not released via releaseState() this should fail or block.
+borrowState returns a Future to access the latest value of the state. If no state is in the storage then assigns it to
+Root.
 */
-func (c *Client) takeState(u update.Update) (state.Handler, error) {
-	handle, isSome := u.StateID().Unwrap()
-	if isSome {
-		if state, found := c.conversationStates[handle]; found {
-			err := state.SetTemplate(c.template)
-			if err != nil {
-				log.Printf("While fetching state from storage for handle %q: %s", handle, err)
-			} else {
-				log.Printf("Lent state %T", state)
-
-				return state, nil
-			}
-		}
+func (c *Client) borrowState(handle string) *borrowonce.Future[state.State] {
+	if future, exists := c.conversationStateStore.Borrow(handle); exists {
+		return future
 	}
 
-	state := &state.Root{}
+	c.conversationStateStore.Set(handle, &state.RootState{})
 
-	if err := state.SetTemplate(c.template); err != nil {
-		return state, fmt.Errorf("while fetching state from storage for the default state: %w", err)
+	if future, exists := c.conversationStateStore.Borrow(handle); exists {
+		return future
 	}
 
-	log.Printf("Lent state %T", state)
-
-	return state, nil
+	// Implementation of borrownonce.Store guarantees that after a value is Set() it is available to borrow
+	panic("conversation store did not lend a value after it was set explicitly")
 }
 
-// TODO: Should unlock the lock that  prevents takeState() from giving out state for the conversation.
-func (c *Client) releaseState(u update.Update, s state.Handler) {
-	if handle, isSome := u.StateID().Unwrap(); !isSome {
-		log.Printf("No state handle for an update %#v", u)
-	} else {
-		log.Printf("Released state: %T", s)
-		c.conversationStates[handle] = s
+/*
+takeState returns a Future to access the latest value of the state. If no state is in the storage then assigns it to
+Root.
+*/
+func (c *Client) borrowUserData(handle update.UserID) *borrowonce.Future[state.UserSharedData] {
+	if future, exists := c.userSharedDataStore.Borrow(handle); exists {
+		return future
 	}
+
+	c.userSharedDataStore.Set(handle, state.UserSharedData{
+		GithubAPIKey: option.None[string](),
+	})
+
+	if future, exists := c.userSharedDataStore.Borrow(handle); exists {
+		return future
+	}
+
+	// Implementation of borrownonce.Store guarantees that after a value is Set() it is available to borrow
+	panic("conversation store did not lend a value after it was set explicitly")
```

Comment:
Same here as in https://github.com/m-kuzmin/daily-reporter/pull/7/files#r1259468099